File: easy/198. House Robber/solution.java

//f(k) = max(robs(k – 2) + nums[i - 1], f(k – 1))
class Solution {
    public int rob(int[] nums) {

        if (nums == null || nums.length == 0) {
            return 0;
        }
        if (nums.length == 1) {
            return nums[0];
        }
        if (nums.length == 2) {
            return Math.max(nums[0], nums[1]);
        }
        int len = nums.length;
        int[] robs = new int[len + 1];
        robs[0] = 0;
        robs[1] = nums[0];
        for (int i = 2; i <= nums.length; i++) {
            robs[i] = Math.max(robs[i - 2] + nums[i - 1], robs[i - 1]);
        }

        int ans = 0;
        for (int rob : robs) {
            if (rob > ans) {
                ans = rob;
            }
        }
        return ans;
    }
}

File: easy/198. House Robber/solution.py
class Solution:
    def rob(self, nums: List[int]) -> int:
        
        if not nums or len(nums) == 0:
            return 0
        dp = [0] * (len(nums) + 1)
        dp[0] = 0
        dp[1] = nums[0]
        for i in range(2, len(nums)+1):
            if nums[i - 1] + dp[i - 2] > dp[i - 1]:
                dp[i] = nums[i - 1] + dp[i - 2]
            else:
                dp[i] = dp[i - 1]
        
        answer = 0
        for each in dp:
            if each > answer:
                answer = each
        return answer